### DISCUSSION TITLE
Sealed classes for states for concurrent state machines

### DISCUSSION THREAD

**Discussion by Lukasmp3** [nsk90/kstatemachine#104] [Q&A] [answered]
I tried the [example](https://github.com/KStateMachine/kstatemachine/blob/master/samples/src/commonMain/kotlin/ru/nsk/samples/MinimalSealedClassesSample.kt) of sealed class for states in concurrently running multiple state machines. They are (shortly living) running separately in own coroutine context. But if I run them concurrently it seems that e.g. the YellowState will have defined multiple SwitchEvent transitions (until one of state machines is destroyed), right? When processing event in such state machine, I get `Multiple transitions match... in YellowState`. I know I can set `doNotThrowOnMultipleTransitionsMatch=true` but if one state machine is destroyed while the other state machine "used" its event I get 'YellowState parent is not set'. I was thinking if the issue might be that my sealed class implements DefaultState for states but sealed class for events implements DataEvent for events (and not just Event)?

**Comment by nsk90** [comment, marked as answer]
I don't recommend using `object` states in multithreading context. Use separate state instances. I suppose this will help.

Events are immutable and are not tied to StateMachine, so you can use them whatever you need.

**Reply by Lukasmp3** [reply]
After the change to separate state instances it seems it's working correctly. Thanks!
